src/parser/ast/expression.rs

```diff
@@ -0,0 +1,215 @@
+use std::fmt::Debug;
+
+use super::DebugSymRef;
+
+#[derive(Clone)]
+pub enum BinaryOperator {
+    Sum,
+    Sub,
+    Mul,
+    Pow,
+    Eq,
+    NEq,
+    And,
+    Or,
+}
+
+impl BinaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use BinaryOperator::*;
+        match self {
+            Eq => true,
+            NEq => true,
+            And => true,
+            Or => true,
+            Sum => false,
+            _ => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for BinaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Sum => write!(f, "+"),
+            Self::Sub => write!(f, "-"),
+            Self::Mul => write!(f, "*"),
+            Self::Pow => write!(f, "^"),
+            Self::Eq => write!(f, "=="),
+            Self::NEq => write!(f, "!="),
+            Self::And => write!(f, "&&"),
+            Self::Or => write!(f, "||"),
+        }
+    }
+}
+
+impl From<String> for BinaryOperator {
+    fn from(value: String) -> Self {
+        use BinaryOperator::*;
+        match value.as_str() {
+            "+" => Sum,
+            "*" => Mul,
+            "^" => Pow,
+            "==" => Eq,
+            "!=" => NEq,
+            "&&" => Or,
+            "||" => And,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum UnaryOperator {
+    Not,
+    Neg,
+}
+
+impl UnaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use UnaryOperator::*;
+        match self {
+            Not => true,
+            Neg => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for UnaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use UnaryOperator::*;
+        match self {
+            Not => write!(f, "!"),
+            Neg => write!(f, "-"),
+        }
+    }
+}
+
+impl From<String> for UnaryOperator {
+    fn from(value: String) -> Self {
+        use UnaryOperator::*;
+        match value.as_str() {
+            "!" => Not,
+            "-" => Neg,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Expression<F, V> {
```

Comment:
How about this?
```rust
pub enum Expression<F, V> {
    BinOp {
        op: BinaryOperator,
        lhs: Box<Node<F, V>>,
        rhs: Box<Node<F, V>>,
    },
    // [...]
}

pub struct Node<F, V>{
  dsym: DebugSymbRef,
  e: Expression<F, V>,
}
```

This way you don't need to add dsym in each possible enum case.

Reply:
But expression is not made or nodes, so we would only get the DebugSymRef in the root, the idea is that we will have the DebugSymRef in each node of the AST.

Reply:
> But expression is not made or nodes, so we would only get the DebugSymRef in the root, the idea is that we will have the DebugSymRef in each node of the AST.

I don't understand what you're saying.  Why do you claim that expression is not made of nodes?  In the snippet I wrote, Expressions are made of nodes, so you get a DebugSYmRef in each node.

Reply:
Ahh, I didn't see the sub expressions are node. I don't know, it will create these two structs that are always including each other, in one way it is more separation of concerns but in other will make the code more difficult perhaps?

Reply:
And alternative would be to give an ID to each node, and have symbol table that references to the DebugSymRef

Reply:
> Ahh, I didn't see the sub expressions are node. I don't know, it will create these two structs that are always including each other, in one way it is more separation of concerns but in other will make the code more difficult perhaps?

I have the intuition that my suggestion may lead to simpler code; but I just wanted to give you the suggestion (not a request).  Feel free to make the choice that seems better to you for the implementation :)

> And alternative would be to give an ID to each node, and have symbol table that references to the DebugSymRef

This would just replace `DebugSymRef` by an `ID` right?  So it doesn't seem an alternative to the suggestion, as it applies in both your original code and my suggestion.

Reply:
Yes, but ID is more flexible because it can lead to different tables with different information. So it is more decoupled